pkg/reboot/rebooter.go

```diff
@@ -27,27 +29,47 @@ func NewWatchdogRebooter(wd watchdog.Watchdog, log logr.Logger) Rebooter {
 	}
 }
 
-func (r *WatchdogRebooter) Reboot() error {
+func (r *WatchdogRebooter) Reboot() (ctrl.Result, error) {
+	if r.isWatchdogRebootStuck() {
+		return r.softwareReboot()
+	}
+	//Watch dog is rebooting, wait to make sure watchdog is rebooting properly otherwise intervene with software reboot
+	if r.wd.IsRebooting() {
+		r.log.Info("watchdog is waiting for reboot to commence, waiting for 10 seconds")
+		return ctrl.Result{RequeueAfter: time.Second * 10}, nil
+	}
+
 	if r.wd == nil || !r.wd.IsStarted() {
 		r.log.Info("no watchdog is present on this host, trying software reboot")
 		//we couldn't init a watchdog so far but requested to be rebooted. we issue a software reboot
-		if err := r.softwareReboot(); err != nil {
-			r.log.Error(err, "failed to run reboot command")
-			// TODO retry because of this?
-			//return err
-			return nil
-		}
-		return nil
+		return r.softwareReboot()
 	}
 	// we stop feeding the watchdog for a reboot
 	r.wd.Stop()
 	r.log.Info("watchdog feeding has stopped, waiting for reboot to commence")
-	return nil
+	return ctrl.Result{RequeueAfter: time.Second * 10}, nil
 }
 
 // softwareReboot performs software reboot by running systemctl reboot
-func (r *WatchdogRebooter) softwareReboot() error {
+func (r *WatchdogRebooter) softwareReboot() (ctrl.Result, error) {
 	// hostPID: true and privileged:true required to run this
 	rebootCmd := exec.Command("/usr/bin/nsenter", "-m/proc/1/ns/mnt", "/bin/systemctl", "reboot", "--force", "--force")
-	return rebootCmd.Run()
+
+	if err := rebootCmd.Run(); err != nil {
+		r.log.Error(err, "failed to run reboot command")
+		// TODO retry because of this?
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *WatchdogRebooter) isWatchdogRebootStuck() bool {
+	timeToAssumeRebootHasStarted := time.Second * 30
+	lastFoodTime := r.wd.LastFoodTime()
+	var isStuck bool
+	now := time.Now()
+	if isStuck = now.After(lastFoodTime.Add(timeToAssumeRebootHasStarted)); isStuck {
+		r.log.Info("watchdog reboot is stuck, about to try software reboot")
```

Comment:
this method should not know what we do when we return it's stuck, the caller should log this IMHO

Reply:
that, and introducing the const, and skipping unneeded new vars can make this a one liner:
`return time.Now().After(r.wd.LastFoodTime().Add(timeToAssumeRebootHasStarted))`

Reply:
>this method should not know what we do when we return it's stuck, the caller should log this IMHO

I agree, I've changed the logging

>that, and introducing the const, and skipping unneeded new vars can make this a one liner:

I've kept some (different) logging so I don't think it's relevant.